File: ios/Runner/FlutterSmartLinksPlugin.swift
import Flutter
import UIKit

/**
 FlutterSmartLinksPlugin (iOS / Swift)

 Handles:
 - Universal Links (HTTPS — requires apple-app-site-association)
 - Custom URI scheme deep links
 - Initial link extraction (cold start)
 - Foreground link stream (EventChannel)
 - Deferred link token persistence (UserDefaults)
 */
public class FlutterSmartLinksPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    // MARK: - Constants

    private static let methodChannelName = "flutter_smart_links"
    private static let eventChannelName  = "flutter_smart_links/events"
    private static let deferredTokenKey  = "flutter_smart_links_deferred_token"

    // MARK: - State

    private var eventSink: FlutterEventSink?
    private var initialLink: String?
    private var pendingLink: String?

    // MARK: - Registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = FlutterSmartLinksPlugin()

        let methodChannel = FlutterMethodChannel(
            name: methodChannelName,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: eventChannelName,
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)

        registrar.addApplicationDelegate(instance)
    }

    // MARK: - FlutterMethodCallDelegate

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getInitialLink":
            result(initialLink)
            initialLink = nil // consume once

        case "storeDeferredToken":
            if let args = call.arguments as? [String: Any],
               let token = args["token"] as? String {
                UserDefaults.standard.set(token, forKey: Self.deferredTokenKey)
            }
            result(nil)

        case "getDeferredToken":
            result(UserDefaults.standard.string(forKey: Self.deferredTokenKey))

        case "clearDeferredToken":
            UserDefaults.standard.removeObject(forKey: Self.deferredTokenKey)
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - FlutterStreamHandler

    public func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        self.eventSink = events
        // Flush any link that arrived before the stream was ready
        if let pending = pendingLink {
            events(pending)
            pendingLink = nil
        }
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    // MARK: - UIApplicationDelegate hooks

    /// Cold-start via custom URI scheme (e.g. myapp://product/123)
    public func application(
        _ app: UIApplication,
        open url: URL,
        options: [UIApplication.OpenURLOptionsKey: Any] = [:]
    ) -> Bool {
        handleIncomingUrl(url.absoluteString)
        return true
    }

    /// Cold-start via Universal Link (HTTPS)
    public func application(
        _ application: UIApplication,
        continue userActivity: NSUserActivity,
        restorationHandler: @escaping ([UIUserActivityRestoring]?) -> Void
    ) -> Bool {
        if userActivity.activityType == NSUserActivityTypeBrowsingWeb,
           let url = userActivity.webpageURL {
            handleIncomingUrl(url.absoluteString)
            return true
        }
        return false
    }

    // MARK: - Private helpers

    private func handleIncomingUrl(_ urlString: String) {
        if eventSink != nil {
            eventSink?(urlString)
        } else if initialLink == nil {
            // First link before stream is ready → treat as initial link
            initialLink = urlString
        } else {
            pendingLink = urlString
        }
    }
}

File: android/src/main/kotlin/ts/sun/flutter_smart_links/FlutterSmartLinksPlugin.kt
package ts.sun.flutter_smart_links

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * FlutterSmartLinksPlugin
 *
 * Handles:
 * - App Links (Android 12+ verified links)
 * - Custom URI scheme deep links
 * - Initial link extraction (cold start)
 * - Foreground link stream (EventChannel)
 * - Deferred link token persistence (SharedPreferences)
 */
class FlutterSmartLinksPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware,
    EventChannel.StreamHandler {

    companion object {
        private const val METHOD_CHANNEL = "flutter_smart_links"
        private const val EVENT_CHANNEL  = "flutter_smart_links/events"
        private const val PREFS_NAME     = "flutter_smart_links_prefs"
        private const val KEY_DEFERRED   = "deferred_token"
    }

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context

    private var eventSink: EventChannel.EventSink? = null
    private var initialLink: String? = null
    private var pendingLink: String? = null

    // ── FlutterPlugin ──────────────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext

        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    // ── ActivityAware ──────────────────────────────────────────────────────

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        initialLink = extractLink(binding.activity.intent)
        binding.addOnNewIntentListener { intent ->
            handleNewIntent(intent)
            true
        }
    }

    override fun onDetachedFromActivityForConfigChanges() {}
    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {}
    override fun onDetachedFromActivity() {}

    // ── MethodCallHandler ──────────────────────────────────────────────────

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getInitialLink" -> {
                result.success(initialLink)
                initialLink = null
            }
            "storeDeferredToken" -> {
                val token = call.argument<String>("token")
                if (token != null) prefs().edit().putString(KEY_DEFERRED, token).apply()
                result.success(null)
            }
            "getDeferredToken" -> result.success(prefs().getString(KEY_DEFERRED, null))
            "clearDeferredToken" -> {
                prefs().edit().remove(KEY_DEFERRED).apply()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    // ── EventChannel.StreamHandler ─────────────────────────────────────────

    override fun onListen(arguments: Any?, sink: EventChannel.EventSink?) {
        eventSink = sink
        pendingLink?.let { link ->
            sink?.success(link)
            pendingLink = null
        }
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // ── Private helpers ────────────────────────────────────────────────────

    private fun handleNewIntent(intent: Intent) {
        val link = extractLink(intent) ?: return
        if (eventSink != null) eventSink?.success(link) else pendingLink = link
    }

    private fun extractLink(intent: Intent?): String? {
        if (intent == null) return null
        val data: Uri? = intent.data
        return if (intent.action == Intent.ACTION_VIEW && data != null) data.toString() else null
    }

    private fun prefs(): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
}
